src/shared_utilities/FileReader.jl

```diff
@@ -0,0 +1,371 @@
+            file_path,
+            varname,
+            surface_space,
+            outfile_root;
+            mono = mono,
+        )
+    end
+    ClimaComms.barrier(comms)
+    data_dates = JLD2.load(
+        joinpath(regrid_dirpath, outfile_root * "_times.jld2"),
+        "times",
+    )
+
+    # init time tracking info
+    current_fields =
+        Fields.zeros(FT, surface_space), Fields.zeros(FT, surface_space)
+    segment_length = [Int(0)]
+
+    # unless the start file date is specified, find the closest one to the start date
+    # TODO if we use two separate data files that start on different dates, we'll need to
+    #  reconcile data_dates[1] between the two of them - maybe add a function for this?
+    date_idx0 =
+        date_idx0 != nothing ? date_idx0 :
+        [
+            argmin(
+                abs.(
+                    parse(FT, datetime_to_strdate(data_dates[1])) .-
+                    parse.(FT, datetime_to_strdate.(data_dates[:]))
+                ),
+            ),
+        ]
+    args = (
+        file_path,
+        comms,
+        varname,
+        data_dates,
+        current_fields,
+        surface_space,
+        deepcopy(date_idx0),
+    )
+    return FileInfo{FT, typeof.(args)...}(
+        args...,
+        date_idx0,
+        segment_length,
+        regrid_dirpath,
+        outfile_root,
+    )
+end
+
+"""
+    read_data_fields!(file_info::FileInfo{FT}, date) where {FT}
+
+Extracts data from regridded (to model grid) NetCDF files.
+The times for which data is extracted depends on the specifications in the
+`file_info` struct).
+Data at one point in time is stored in `file_info.data_fields[1]`, and
+data at the next time is stored in `file_info.data_fields[2]`. With these
+two data fields saved, we can interpolate between them for any dates
+in this range of time.
+
+# Arguments
+- `date`: [Dates.DateTime] start date for data.
+- `file_info`: [FileInfo] containing boundary condition data.
+"""
+function read_data_fields!(file_info::FileInfo{FT}, date) where {FT}
+    (; comms, regrid_dirpath, varname, all_dates, outfile_root) = file_info
+    date_idx = file_info.date_idx[1]
+    date_idx0 = file_info.date_idx0[1]
+
+    # Case 1: Current date is before or at first date in data file
+    #  Load in data at first date for both `data_fields[1]` and `data_fields[2]`
+    if (date_idx == date_idx0) && (Dates.days(date - all_dates[date_idx]) <= 0)
+        if date !== all_dates[date_idx]
+            @warn "this time period is before input data - using file from $(all_dates[date_idx0])"
+        end
+
+        file_info.data_fields[1] .= Regridder.swap_space!(
+            Regridder.read_from_hdf5(
+                regrid_dirpath,
+                outfile_root,
+                all_dates[Int(date_idx0)],
+                varname,
+                comms,
+            ),
+            file_info.surface_space,
+        )
+        file_info.data_fields[2] .= deepcopy(file_info.data_fields[1])
```

Comment:
Why are we calling `deepcopy` here?

Reply:
In these cases, we want `data_fields[1]` and `data_fields[2]` to store exactly the same information. Would it use less space to just call `copy` instead?

Reply:
`file_info.data_fields[2] .= file_info.data_fields[1]` should be fine. `file_info.data_fields[2] .= deepcopy(file_info.data_fields[1])` first allocates a whole new copy of `file_info.data_fields[1]`, and then values from that array get broadcasted into `file_info.data_fields[2]`

Reply:
I've removed the calls to `deepcopy` and am just assigning `data_fields` using `.=`. This is clearer and has fewer allocations